Source/Xamarin/Prism.DryIoc.Forms/PrismApplication.cs

```diff
@@ -128,7 +128,7 @@ protected override void ConfigureViewModelLocator()
         /// <param name="view"><see cref="Page" /> navigated to</param>
         /// <param name="type"><see cref="Type" /> to resolve</param>
         /// <param name="navigationService">Overriding instance of <see cref="INavigationService" /></param>
-        protected virtual void ResolveTypeForPage(Page view, Type type, INavigationService navigationService)
+        protected internal virtual void ResolveTypeForPage(Page view, Type type, INavigationService navigationService)
```

Comment:
I like that - it makes it clear that the navigation service utilize the page instance. Also I was thinking refactoring ConfigureViewModel like this

````c#
protected override void ConfigureViewModelLocator()
        {
            ViewModelLocationProvider.SetDefaultViewModelFactory((view, type) =>
            {
                var page = view as Page;
                if (page != null)
                {
                    return PageViewModelFactory(page, type);
                }
                return Container.Resolve(type);
            });
        }

        /// <summary>
        /// Resolve the view model of <paramref name="type"/> associated with <paramref name="page"/>
        /// </summary>
        /// <remarks>
        /// The method will set the <see cref="IPageAware.Page" /> property on the <see cref="INavigationService"/>  
        /// instance that will be injected into the view model.
        /// </remarks>
        /// <param name="page">The <see cref="Page"/> associated with the view model</param>
        /// <param name="type">View model type to resolve</param>
        /// <returns>View model instance of type <paramref name="type"/></returns>
        protected virtual object PageViewModelFactory(Page page, Type type)
        {
            // Use CreateNavigationServiceForPage here instead
            var navigationService = CreateNavigationService();
            ((IPageAware)navigationService).Page = page;
            ResolveTypeForPage(page, type, navigationService);
            // Resolve type using the instance navigationService
            var resolver = Container.Resolve<Func<INavigationService, object>>(type);
            return resolver(navigationService);
        }
````

Moving back to the original question, yes, it feels dirty. That method exists soley for the purpose of grabbing the instance injected into the view model so that it can be tested that the `Page` property of `IPageAware` is correct.

Another option would be to add extension methods (for test only) to `INavigationService` that returns the view model navigated to. E.g. `Task<TViewModel> NavigateTo<TViewModel>()`. Then we can check that the navigation service in the mock view model is correctly initialized.

If you have a better way please share